### DISCUSSION TITLE
LorentzVector vs. TLorentzVector

### DISCUSSION THREAD

**Discussion by naidoo88** [dglazier/chanser#21] [User problems] [answered]
I'm hitting the snag that things are LorentzVectors rather than TLorentzVectors, and so I can't do quick operations like .Perp(), .Angle() etc.

Is there a neat/tidy way to convert from LV to TLV?  For some reason it doesn't seem like LV can be cast to a TLV (although I really don't understand why...) 

Cheers!

**Comment by dglazier** [comment, marked as answer]
TLorentzVector is slow and cumbersome and not optimally suited to large volume data analysis. LorentzVectors are lighter and significantly faster. They are two completely different classes, so there is no inheritance or casting between them. You can usually get the same information from LorentzVectors using their utility functions

https://root.cern/doc/v612/Vector.html for overview

https://root.cern/doc/v612/namespaceROOT_1_1Math_1_1VectorUtil.html for helper functions.

So you might want to add (in the function where you use angle)

using ROOT::Math::VectorUtil::Angle;

using ROOT::Math::VectorUtil::Perp;

then

auto angleDiff= Angle(p1,p2);

auto perpComp= Perp(p1,p2);

You can also look at https://github.com/dglazier/chanser/blob/master/core/Kinematics.h

For more examples of LorentzVector manipulation. I think it would also be a good idea to include wrapper functions to Angle and Perp as part of the Kinematics class  to simplify this.

i.e. it would just be

auto angleDiff= kin.Angle(p1,p2); 

I will start a discussion in developments for what we can add to Kinematics

**Reply by naidoo88** [reply]
Thanks Derek!

Works a treat for Angle, but I think the VectorUtil::Perp is doing something different to TLorentzVector::Perp which simply returns the transverse component of the vector in terms  of cylindrical coordinates.

I can of course just define a function which returns sqrt(X*X + Y*Y) - the question then becomes, where is the best place to put user-defined functions like this in a chanser project?

**Reply by dglazier** [reply]
> I can of  course just define a function which returns sqrt(XX + YY) - the question then becomes, where is the best place to put user-defined functions like this in a chanser project?

And to answer this, if it is generally useful functions we should put them in Kinematics. If it is specific to your analysis, or you are developing it then you can make the function a member of your finalstate class, or add as a utility function in your final state header file if it does not need access to any class data. Is that clear ?

**Reply by naidoo88** [reply]
I believe so..  I'll know for sure when I cross the bridge of having to do it!  If I come against any snags I will post up again.

**Comment by dglazier** [comment]
Ah, my mistake, that is the perp component of 1 relative to another. There should be a Perp2 function in LorentzVector itself (leaving you to do the sqrt only if necessary) see
https://root.cern/doc/v612/classROOT_1_1Math_1_1LorentzVector.html#a37d4e60103b9a93548341f45f4930a73

auto perpComp=TMath::Sqrt(p4.Perp2());

**Reply by naidoo88** [reply]
So there is, cracking - thanks!!
